### PR TITLE
Add and Remove parties

### PR DESCRIPTION
**Pull request by tudorcebere**

### CODE REVIEW

**Comment by elichai** [comment]
The new `simulate_dkr_add` test fails

**Comment by elichai** [comment]
Another thing, IIUC the approach you talk is as follows:
t parties calculate their additive shares, they then do PVSS with the additive share in `f_i(0)`(such that `f(0)=Σshareᵢ`) and they generate `n+1` that everyone sums up.
this requires all n parties to be online to recieve their points, sum them, and verify the commitments.

Instead we could do the following:

Notation: `πᵢ(x)` is the polynomial representing the Lagrange Coefficient required for interpolation (`(x-x₁)(x-x₂)/(x₁-x₂)(x₁-x₃)....`)

So t parties come together, and they calculate:
`vᵢ = πᵢ(n+1)*yᵢ`
Then use SSS and generate polynomials `gᵢ(x)` with `gᵢ(0) = vᵢ`, and use PVSS to compute together a polynomial `g(x)` of degree `t-1` where `g(0)=Σvᵢ`.
Then each party encrypts their point using P_n+1's paillier key and broadcast it, Which then decrypts them and interpolates `g(0)` to find their point.
Now party `n+1` has the point `f(n+1)` and no party needs to modify their own secret.
(note that P_n+1 will need to check that against the commitments of the PVSS etc. to make sure their not being lied to)

File: src/lib.rs [hunk 1]
```diff
@@ -31,9 +36,37 @@ pub struct RefreshMessage<P> {
...
+    new_party: bool,
 }
 
 impl<P> RefreshMessage<P> {
+    pub fn distribute_new_party(party_index: usize, t: usize, n: usize) -> (Self, Keys)
+    where
+        P: ECPoint + Clone + Zeroize,
+        P::Scalar: PartialEq + Clone + Debug + Zeroize,
+    {
+        let new_party_key = Keys::create(0);
```
[the remark this answers is not in the document]
Comment (tudorcebere) [reply src/lib.rs:48]: Sorry for the late response, I wasn't available in the past week. I will address the review asap.

[the remark this answers is not in the document]
Comment (tudorcebere) [reply src/lib.rs:48]: Basically, at this point, the new parties does not know his index (I will fix the function signature asap). The other parties needs to decide on their index, it's not available at this stage.

File: src/lib.rs [hunk 2]
```diff
@@ -152,67 +191,154 @@ impl<P> RefreshMessage<P> {
...
-                    Y: refresh_message.points_committed_vec[i].clone(),
-                };
-                if refresh_message.fairness_proof_vec[i]
-                    .verify(&statement)
-                    .is_err()
-                {
-                    return Err(FsDkrError::FairnessProof);
-                }
-            }
-        }
```
[the remark this answers is not in the document]
Comment (tudorcebere) [reply src/lib.rs:172]: Currently, this is the only part failing in the check, I will understand why it fails and add it back.

File: src/test.rs [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+                RefreshMessage::distribute_new_party(party_index, t, n);
+
+            let mut broadcast_vec = Vec::new();
+            let mut new_dks = Vec::new();
+
+            // TODO verify new key
+            for key in keys.iter_mut() {
+                // add the new ek to each party
+                key.paillier_key_vec[party_index - 1] = new_party_refresh_message.ek.clone();
+                let (refresh_message, new_dk) = RefreshMessage::distribute(key);
```
[the remark this answers is not in the document]
Comment (tudorcebere) [reply src/test.rs:106]: Hmm, I think you are right.

**Comment by tudorcebere** [comment]
I need to take this on paper, if you have any links to check on this, it would be very useful!

**Pull request merged**